### PR TITLE
Feature/spectral granger

### PR DESCRIPTION
**Pull request by ackurth**
Implementation of non-parametric spectral Granger Causality.

### CODE REVIEW

**Comment by Moritz-Alexander-Kern** [comment]
Hi @ackurth @rjurkus , (please let me know if further authors should be considered for this PR)

Thanks for this nice addition to elephant :rocket: .

@morales-gregorio kindly offered to give a review for this PR :1st_place_medal: .

------
**~~Bug: readthedocs build, nbsphinx** ~~
~~It seems there is an Issue with the readthedocs build, that is related to the asset tutorial notebook and the release of nbsphinx 0.9.0.~~

**Update 14.03.2023:** This seems to be fixed with [nbsphinx release 0.9.1](https://github.com/spatialaudio/nbsphinx/releases/tag/0.9.1)

**Comment by morales-gregorio** [comment]
Hey! 

I just checked the coverage of the tests and a few edge cases (mostly raising errors and dealing with different input types) are never raised by the tests. Have a look at the report (which you can re-generate using the [coverage package)](https://coverage.readthedocs.io/en/7.2.1/) ) 

The report, an html which you can open with the browser: [Coverage for causality_granger.py 96%.zip](https://github.com/NeuralEnsemble/elephant/files/10978678/Coverage.for.causality_granger.py.96.zip)

File: elephant/causality/granger.py [hunk 1]
```diff
@@ -596,3 +776,163 @@ def conditional_granger(signals, max_order, information_criterion='aic'):
...
+
+    # Calculate cross spectrum for signals
+    freqs, S = segmented_multitaper_cross_spectrum(
+        signals=signals, n_segments=n_segments, len_segment=len_segment,
+        frequency_resolution=frequency_resolution, overlap=overlap, fs=fs,
+        nw=nw, num_tapers=num_tapers, peak_resolution=peak_resolution,
+        return_onesided=False)
+
+    # Remove units attached by the multitaper_cross_spectrum
+    if isinstance(S, pq.Quantity):
```
Comment (morales-gregorio) [inline elephant/causality/granger.py:890]: Why just ignore the units? Maybe you can ensure they are always rescaled to a certain value?

File: elephant/causality/granger.py [hunk 2]
```diff
@@ -596,3 +776,163 @@ def conditional_granger(signals, max_order, information_criterion='aic'):
...
+
+    total_interdependence = (directional_causality_x_y
+                             + directional_causality_y_x
+                             + instantaneous_causality)
+
+    spectral_causality = Causality(
+        directional_causality_x_y=directional_causality_x_y,
+        directional_causality_y_x=directional_causality_y_x,
+        instantaneous_causality=instantaneous_causality,
+        total_interdependence=total_interdependence)
```
Comment (morales-gregorio) [inline elephant/causality/granger.py:936]: Is there a particular reason why to make this into such an object? I have not seen such an output anywhere else, it might confuse some people when using it

**Review by morales-gregorio — CHANGES_REQUESTED** [review]
I just noted a few issues here and there. I will double check the formulae later on

File: elephant/causality/granger.py [hunk 3]
```diff
@@ -596,3 +776,163 @@ def conditional_granger(signals, max_order, information_criterion='aic'):
...
+        Default: 1.0
+    nw : float, optional
+        Time bandwidth product
+        Default: 4.0
+    num_tapers : int, optional
+        Number of tapers used in 1. to obtain estimate of PSD. By default,
+        [2*nw] - 1 is chosen.
+        Default: None
+    peak_resolution : pq.Quantity float, optional
+        Quantity in Hz determining the number of tapers used for analysis.
```
[the remark this answers is not in the document]
Comment (ackurth) [reply elephant/causality/granger.py:825]: This  is spelled out in detail in the documentation `segmented_multitaper_cross_spectrum`, to which we link in the first bulletpoint in the documentation of the function.
The reason is to avoid a duplication.
Do you think it is necessary here?

File: elephant/causality/granger.py [hunk 4]
```diff
@@ -357,6 +381,162 @@ def _optimal_vector_arm(signals, dimension, max_order,
...
+        Number of time series
+
+    Returns
+    -------
+    causal_part : np.ndarray
+        Causal part of cross-spectrum of multivariate time series
+    """
+    # Get coefficients from spectrum
+    causal_part = np.fft.ifft(spectrum, axis=0)
+    # Version 1
```
[the remark this answers is not in the document]
Comment (ackurth) [reply elephant/causality/granger.py]: Addressed in 5e2627d

File: elephant/causality/granger.py [hunk 5]
```diff
@@ -357,6 +381,162 @@ def _optimal_vector_arm(signals, dimension, max_order,
...
+    factorization and terminates if either the maximum number of iterations is
+    reached or the difference between the cross spectrum and the approximate
+    cross spectrum calculated from the covariance matrix and transfer function
+    is sufficiently small.
+
+    Parameters
+    ----------
+    cross_spectrum : np.ndarray
+        Cross spectrum to be decomposed in covariance matrix and transfer
+        function
```
[the remark this answers is not in the document]
Comment (ackurth) [reply elephant/causality/granger.py:466]: Addressed in 5e2627d

File: elephant/causality/granger.py [hunk 6]
```diff
@@ -357,6 +381,162 @@ def _optimal_vector_arm(signals, dimension, max_order,
...
+        Y = np.linalg.solve(factorization,
+                            _dagger(X))
+        Y += identity
+        Y = _bracket_operator(Y, num_freqs, num_signals)
+
+        factorization = np.matmul(factorization, Y)
+
+        diff = factorization - factorization_old
+        error = np.max(np.abs(diff))
+        if error < 1e-12:
```
[the remark this answers is not in the document]
Comment (ackurth) [reply elephant/causality/granger.py]: Addressed in c9e6517.

File: elephant/causality/granger.py (hunk 2 above)
Comment (morales-gregorio) [reply elephant/causality/granger.py:947]: As discussed in person, agree to keep this

File: elephant/causality/granger.py [hunk 7]
```diff
@@ -14,18 +14,30 @@
...
+basics the multitaper method to estimate power- and cross-spectra (e.g.
+sampling frequency, DPSS, time-half bandwidth product).
 
 Implementation
 --------------
 The mathematical implementation of Granger causality methods in this module
 closely follows :cite:`granger-Ding06_0608035`.
+The implementation of spectral Granger causality follows
+:cite:`granger-Dhamala02_354`, :cite:`granger-Wend13_20110610` and
+:cite:`granger-Wilson72` for the spectral matrix factorization.
```
Comment (morales-gregorio) [inline elephant/causality/granger.py:40]: These references do not seem to have been included to the bibtex [elephant.bib](https://github.com/NeuralEnsemble/elephant/blob/03275d39c00acd60eaa21472ec2b0851f59cbbbe/doc/bib/elephant.bib)

Comment (ackurth) [reply elephant/causality/granger.py]: Addressed in [d1f0dc8](https://github.com/NeuralEnsemble/elephant/pull/545/commits/d1f0dc8094012587c9784ffc275c270ed492a3bc).

**Comment by ackurth** [comment]
The coverage was taken care of in 5712bf4 .

File: elephant/causality/granger.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (morales-gregorio) [reply elephant/causality/granger.py:832]: I see this was already cross referenced earlier in the docstring

File: elephant/causality/granger.py (hunk 7 above)
Comment (morales-gregorio) [reply elephant/causality/granger.py]: Great!

File: elephant/causality/granger.py [hunk 8]
```diff
@@ -596,3 +782,168 @@ def conditional_granger(signals, max_order, information_criterion='aic'):
...
+    The spectral Granger Causality is obtained through the following steps:
+
+    1. Determine the cross spectrum of the two signals by applying
+       :func:`segmented_multitaper_cross_spectrum` to the joint signal. See the
+       documentation of this function for the parameter hierarchy of the
+       used for the estimation of the cross spectrum.
+
+    2. Calculate the spectral factorization of the cross spectrum decomposing
+       the very same thing into the covariance matrix and the transfer
+       function.
```
Comment (Junji110) [inline elephant/causality/granger.py:803]: Not clear what "the very same thing" means.

File: elephant/causality/granger.py [hunk 9]
```diff
@@ -596,3 +782,168 @@ def conditional_granger(signals, max_order, information_criterion='aic'):
...
+
+    3. Calculate the directional and instantaneous spectral Granger Causality
+       from the power spectra, the transfer function and the covariance matrix.
+
```
Comment (Junji110) [inline elephant/causality/granger.py:807]: Reference to relevant literature would be preferable.

File: elephant/test/test_causality.py [hunk 10]
```diff
@@ -285,5 +286,223 @@ def test_non_zero_conditional_causality(self):
             self.non_zero_signal, 10, 'bic'), 0.0)
 
 
+class PairwiseSpectralGrangerTestCase(unittest.TestCase):
+
+    def setUp(self):
+        pass
+
+    @staticmethod
+    def _generate_ground_truth_spectral(length_2d=300000, return_coeffs=False,
```
Comment (Junji110) [inline elephant/test/test_causality.py:315]: I think this function should take the weights as a parameter, rather than hard-coded within the function.

File: elephant/causality/granger.py [hunk 11]
```diff
@@ -357,6 +381,168 @@ def _optimal_vector_arm(signals, dimension, max_order,
...
+    -------
+    causal_part : np.ndarray
+        Causal part of cross-spectrum of multivariate time series
+    """
+    # Get coefficients from spectrum
+    causal_part = np.fft.ifft(spectrum, axis=0)
+    # Throw away of acausal part
+    causal_part[(num_freqs + 1) // 2:] = 0
+
+    # Treat zero frequency part
```
Comment (Junji110) [inline elephant/causality/granger.py:408]: Confused by "zero frequency" as we are now working in the time domain...?

**Review by Junji110 — CHANGES_REQUESTED** [review]
Made some comments in the code.

File: elephant/causality/granger.py [hunk 12]
```diff
@@ -596,3 +782,168 @@ def conditional_granger(signals, max_order, information_criterion='aic'):
...
+                              overlap=0.5, num_iterations=300,
+                              term_crit=1e-12):
+    r"""Determine spectral Granger Causality of two signals.
+
+    The spectral Granger Causality is obtained through the following steps:
+
+    1. Determine the cross spectrum of the two signals by applying
+       :func:`segmented_multitaper_cross_spectrum` to the joint signal. See the
+       documentation of this function for the parameter hierarchy of the
+       used for the estimation of the cross spectrum.
```
[the remark this answers is not in the document]
Comment (ackurth) [reply elephant/causality/granger.py]: Addressed in 0f6ea81.

File: elephant/causality/granger.py (hunk 8 above)
Comment (ackurth) [reply elephant/causality/granger.py]: Addressed in 0f6ea81.

File: elephant/causality/granger.py [hunk 13]
```diff
@@ -596,3 +782,168 @@ def conditional_granger(signals, max_order, information_criterion='aic'):
...
+        Default: 1e-12
+
+    Returns
+    -------
+    freqs : np.ndarray
+        Frequencies associated with the spectral Granger Causality estimate.
+    Causality
+        A `namedtuple` with the following attributes:
+            directional_causality_x_y : np.ndarray
+                Spectrally resolved Granger causality of X influence onto Y.
```
[the remark this answers is not in the document]
Comment (ackurth) [reply elephant/causality/granger.py]: Addressed in 0f6ea81.

File: elephant/causality/granger.py (hunk 9 above)
Comment (ackurth) [reply elephant/causality/granger.py:807]: Addressed in 0f6ea81.

File: elephant/causality/granger.py [hunk 14]
```diff
@@ -596,3 +782,168 @@ def conditional_granger(signals, max_order, information_criterion='aic'):
...
+    S = np.transpose(S, axes=(1, 0, 2))
+
+    # Decompose cross spectrum into covariance and transfer function
+    C, H = _spectral_factorization(S, num_iterations=num_iterations)
+
+    # Take positive frequencies
+    mask = (freqs >= 0)
+    freqs = freqs[mask]
+
+    S = np.transpose(S, axes=(2, 0, 1))[mask]
```
[the remark this answers is not in the document]
Comment (ackurth) [reply elephant/causality/granger.py]: Addressed in [2d88ed4](https://github.com/NeuralEnsemble/elephant/pull/545/commits/2d88ed40ed32887170bab4616465be5bbdad5021).

File: elephant/test/test_causality.py (hunk 10 above)
Comment (ackurth) [reply elephant/test/test_causality.py]: Addressed in [2d88ed4](https://github.com/NeuralEnsemble/elephant/pull/545/commits/2d88ed40ed32887170bab4616465be5bbdad5021).

File: elephant/causality/granger.py (hunk 11 above)
Comment (ackurth) [reply elephant/causality/granger.py]: Nice catch.
Addressed in [2d88ed4](https://github.com/NeuralEnsemble/elephant/pull/545/commits/2d88ed40ed32887170bab4616465be5bbdad5021).

**Comment by morales-gregorio** [comment]
From my side the changes look good to me

**Comment by Moritz-Alexander-Kern** [comment]
Many thanks to @Junji110 and @morales-gregorio  for the reviews. :trophy: 

I hope I didn't get this wrong, but @rjurkus is intending to add some validation tests to this PR (?).
Please let me know if I'm confusing this and actually this will be a separate PR.

File: elephant/causality/granger.py [hunk 15]
```diff
@@ -596,3 +781,173 @@ def conditional_granger(signals, max_order, information_criterion='aic'):
...
+    # Transpose cross spectrum due to different conventions used in
+    # segemented_multitaper_cross_spectrum and the calculations of spectral
+    # Granger causality
+    S = np.transpose(S, axes=(1, 0, 2))
+
+    # Move frequencies to first axis - Needed for _spectral_factorization
+    S = np.transpose(S, axes=(2, 0, 1))
+
+    # Decompose cross spectrum into covariance and transfer function
+    C, H = _spectral_factorization(S, num_iterations=num_iterations)
```
Comment (Moritz-Alexander-Kern) [inline elephant/causality/granger.py:914]: [Optional, Question] 
Why is the parameter `term_crit` of `pairwise_spectral_granger` not passed as an argument to `_spectral_factorization` ?, like:
```python
    C, H = _spectral_factorization(S, num_iterations=num_iterations, term_crit=term_crit)
```

If I get this correctly `_spectral_factorization` will always use the default value `1e-12` , is this intended? (independent of `term_crit` passed to `pairwise_spectral_granger`)

**Review by Moritz-Alexander-Kern — APPROVED** [review]
Thanks for this nice addition to Elephant @ackurth and @rjurkus .

My comments are mostly related to formatting of the docs and can be ignored. 
I am intending to include them in #553 .

One question came up, however, this could only be lack of understanding on my part. 
Just wanted to make sure you are aware. 

Please ping me if I can merge.

**Comment by rjurkus** [comment]
> I hope I didn't get this wrong, but @rjurkus is intending to add some validation tests to this PR (?). Please let me know if I'm confusing this and actually this will be a separate PR.

I will add a couple of tests once the `elphant-data` PR is merged.

**Review by Moritz-Alexander-Kern — COMMENTED** [review]
build wheels
